### PR TITLE
DOC Document ORM changes for CMS 5.2

### PR DESCRIPTION
**Pull request by GuySartorelli**
## This PR includes
- [x] Docs for UNION
- [x] Docs for WITH
- [x] Docs for `DataQuery::setAllowCollidingFieldStatements()` (in changelog only, but feel free to point out a good place for it if you think it deserves its own entry in the main docs somewhere)
- [x] Update docs that refer to joins to include RIGHT JOIN
- [x] Changelog entry

Note that the documentation here intentionally doesn't cover every scenario - the `08_SQL_Select.md` file already says that there's a lot in here and that people should check the PHPDoc for full usage details.

## Issue
- https://github.com/silverstripe/silverstripe-framework/issues/10902

### CODE REVIEW

File: en/04_Changelogs/5.2.0.md
Comment (ChloeSartorelli) [inline en/04_Changelogs/5.2.0.md:94]: Gives a unique anchor ID for this heading to distinguish it from the `silverstripe/framework` heading that will be automatically generated in the actual changelog below this.
Not strictly necessary as part of this PR but it's a tiny change so I probably wouldn't bother with a separate PR.

File: en/02_Developer_Guides/00_Model/01_Data_Model_and_ORM.md
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/00_Model/01_Data_Model_and_ORM.md:640]: I'm intentionally cramming the common three ways to refer to this functionality into the heading for search discovery purposes:
- "Common Table Expressions"
- "CTE"
- "WITH clause"

File: en/02_Developer_Guides/00_Model/08_SQL_Select.md
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/00_Model/08_SQL_Select.md:109]: I've intentionally not added any more documentation than this for unions because I think the method is self-explanatory for anyone who knows what a union is and how it's used in SQL.

Frankly, anyone who doesn't understand the SQL concept of unions probably won't care about it anyway or have any reason to use it.

This is also roughly the same amount of documentation available for joins, which are arguably more complex to use as far as the API is concerned.

**Comment by emteknetnz** [comment]
I left a review on https://github.com/silverstripe/silverstripe-framework/pull/10943 - depending on that we may want to change stuff in this PR

**Comment by GuySartorelli** [comment]
Rebased to resolve conflicts

**Comment by GuySartorelli** [comment]
There have been some changes in the code, but I'll hold off updating docs until that PR has been approved.

**Comment by GuySartorelli** [comment]
Updated docs to reflect the recent code changes, and to make some things slightly clearer.

**Review by emteknetnz — APPROVED** [review]

**Pull request merged**
